Adding "start clearing range" to the Obstacle Layer.

adrienbarral opened the pull request:
Start clearing range is the minimal distance at which a sensor can see an obstacle. This is useful for all stereo based sensor (like RGBD and stereo cameras). Obstacles  will not be cleared between the sensor and this range.

I code this patch because I encounter the follwing problem. I have a mobile robot with an RGBD sensor moving in a place where there is a lot of shelfs (or low table). When the robot is far of this objects, there is no problem, they are seend as obstacles. But if the robot go close of this objects, they are cleared as the robot move...

If you want further explanation ask me, I can create a video With and Without my patch.

Commit:
Adding "start clearing range" to the Obstacle Layer. Start clearing range is the minimal distance at which a sensor can see an obstacle. This is useful for all stereo based sensor (like RGBD and stereo cameras). Obstacles  will not be cleared between the sensor and this range.

mikeferguson commented:
This is half of https://github.com/ros-planning/navigation/issues/124

romainreignier commented on costmap_2d/include/costmap_2d/observation_buffer.h:
```diff
@@ -70,15 +70,16 @@ class ObservationBuffer
    * @param  max_obstacle_height The minimum height of a hitpoint to be considered legal
    * @param  obstacle_range The range to which the sensor should be trusted for inserting obstacles
    * @param  raytrace_range The range to which the sensor should be trusted for raytracing to clear out space
+   * @param  start_clearing_range The range to wich the sensor start to be trusted for clearing out space.
```
"to wHich"

adrienbarral commented:
Yes this look like to be the half of #124. But my patch works for all data source (not only LaserScan). I think this is important to introduce a new parameter in the source properties rather than use the min_rage data of LaserScan. Because, using this new parameter "start_clearing_range", we can work with all data sources (not only LaserScan).

procopiostein commented:
I agree that your implementation is more generic than of  #124.
But I would propose a different nomeclature, to keep some consistency with the other parameters, for example, naming it "min_raytrace_range" which defaults to 0. 
Also there could be another parameter "min_obstacle_range" that would do the other half of #124, controlling from what distance the obstacles should be marked.

adrienbarral commented:
Ok, I can rename my parameter to min_raytrace_range. And I will implement the second half of #124 with this "min_obstacle_range".
To be sure that I well understood what "min_obstacle_range" must do. If we have an hit (a point sensed) at a range of 2 meters and that min_obstacle_range is 1.5m, this point must be discarded (no clearing, and no point set). Isn't it  ?
Or should I clear between min_raytrace_range and min_obstacle_range, and not set the point as a lethal obstacle ? Or something else ?

I think I can post seomething before the end of this week.

procopiostein commented:
> To be sure that I well understood what "min_obstacle_range" must do. If we have an hit (a point sensed) at a range of 2 meters and that min_obstacle_range is 1.5m, this point must be discarded (no clearing, and no point set). Isn't it ?

Yes, that how I would interpret  #124 second part in a more generic form.
Anyway, maybe you should wait for some feedback from @DLu and @mikeferguson on that.

thanks!

mikeferguson commented on costmap_2d/include/costmap_2d/observation.h:
```diff
@@ -51,7 +51,7 @@ class Observation
    */
   Observation() :
-    cloud_(new pcl::PointCloud<pcl::PointXYZ>()), obstacle_range_(0.0), raytrace_range_(0.0)
+    cloud_(new pcl::PointCloud<pcl::PointXYZ>()), obstacle_range_(0.0), raytrace_range_(0.0), start_clearing_range_(0.0)
```
Let's call this new parameter "min_raytrace_range_". Let's also update raytrace_range_ to be max_raytrace_range_

mikeferguson commented on costmap_2d/include/costmap_2d/observation.h:
```diff
@@ -68,9 +68,9 @@ class Observation
+              double obstacle_range, double raytrace_range, double start_clearing_range=0.0) :
       origin_(origin), cloud_(new pcl::PointCloud<pcl::PointXYZ>(cloud)),
-      obstacle_range_(obstacle_range), raytrace_range_(raytrace_range)
+      obstacle_range_(obstacle_range), raytrace_range_(raytrace_range),start_clearing_range_(start_clearing_range)
```
let's leave this function signature without change (you won't be able to set min through it), but default min_raytrace_range_ to 0.0, and add a deprecation on this constructor. Let's then create a new constructor which has min_range BEFORE max_range.

This way, code people have will still compile (albeit with a deprecation warning) but new code that sets mins will look more "normal" by having the min parameter before the max parameter in the parameter list.

mikeferguson commented on costmap_2d/include/costmap_2d/observation.h:
```diff
@@ -90,13 +91,16 @@ class Observation
   pcl::PointCloud<pcl::PointXYZ>* cloud_;
   double obstacle_range_, raytrace_range_;
+  double start_clearing_range_;
+
```
remove extra newline, uneeded change

mikeferguson commented on costmap_2d/include/costmap_2d/observation_buffer.h:
```diff
@@ -70,15 +70,16 @@ class ObservationBuffer
-                    double raytrace_range, tf::TransformListener& tf, std::string global_frame,
-                    std::string sensor_frame, double tf_tolerance);
+                    double raytrace_range, double start_clearing_range, tf::TransformListener& tf,
+                    std::string global_frame, std::string sensor_frame, double tf_tolerance);
```
same comments as we did for Observation -- leave this constructer, default min_raytrace_range to 0.0, deprecate constructor. Add new constructor with min before max. Since we'll rename raytrace range to max_raytrace_range, let's also make sure the doxygen gets updated

mikeferguson commented on costmap_2d/include/costmap_2d/observation_buffer.h:
```diff
@@ -159,6 +160,7 @@ class ObservationBuffer
   boost::recursive_mutex lock_;  ///< @brief A lock for accessing data in callbacks safely
   double obstacle_range_, raytrace_range_;
   double tf_tolerance_;
+  double start_clearing_range_;
```
reaname to min_raytrace_range_, update raytrace_range to be max_raytrace_range_, and let's declare all of them on the same line above: "double obstacle_range_, min_raytrace_range_, max_raytrace_range_".

mikeferguson commented on costmap_2d/plugins/obstacle_layer.cpp:
```diff
@@ -114,7 +114,7 @@ void ObstacleLayer::onInitialize()
     }
 
-    std::string raytrace_range_param_name, obstacle_range_param_name;
+    std::string raytrace_range_param_name, obstacle_range_param_name, start_clearing_range_param_name;
```
i would update these to be min/max_raytrace_range_param_name

mikeferguson commented on costmap_2d/plugins/obstacle_layer.cpp:
```diff
@@ -130,15 +130,21 @@ void ObstacleLayer::onInitialize()
     }
 
+    double start_clearing_range = 0.0;
+    if (source_node.searchParam("start_clearing_range", start_clearing_range_param_name))
```
min_raytrace_range

mikeferguson commented on costmap_2d/plugins/obstacle_layer.cpp:
```diff
@@ -130,15 +130,21 @@ void ObstacleLayer::onInitialize()
       source_node.getParam(raytrace_range_param_name, raytrace_range);
```
We should update this to search for max_raytrace_range, if not found, it looks for raytrace_range, but prints a warning that raytrace_range is deprecated (and yes, this starts to leak into the config for each robot, but we will give several/many release cycles before we remove the deprecated part)

mikeferguson commented on costmap_2d/plugins/obstacle_layer.cpp:
```diff
@@ -525,6 +531,31 @@ void ObstacleLayer::raytraceFreespace(const Observation& clearing_observation, d
   {
     double wx = cloud.points[i].x;
     double wy = cloud.points[i].y;
+    if(clearing_observation.start_clearing_range_ > 0)
```
space after if

mikeferguson commented on costmap_2d/plugins/obstacle_layer.cpp:
```diff
@@ -525,6 +531,31 @@ void ObstacleLayer::raytraceFreespace(const Observation& clearing_observation, d
+      // If this sensor start seeing obstacle at a certain range, we must move x0,y0 at this
+      // range from the sensor (following the line going from the sensor to the point)
+      double obstacle_range = hypot(wx-ox, wy-oy);
+      if(obstacle_range < clearing_observation.start_clearing_range_)
```
space after if

mikeferguson commented on costmap_2d/plugins/obstacle_layer.cpp:
```diff
@@ -525,6 +531,31 @@ void ObstacleLayer::raytraceFreespace(const Observation& clearing_observation, d
+        // In this particular case, we just don't set free spaces (after displaying a warning).
+        ROS_WARN_THROTTLE(
+                1.0, "Your start_clearing_range parameter seems to be too short because an observation was returned by"
+                "your sensor in this range. Cells between sensor and this observation was not cleared");
```
I'm not a fan of this log message -- there are certainly sensors that report 0 when they cannot see something, and this would spam the crap out of the console/logs. I would suggest we either remove this, or make it a ROS_DEBUG_NAMED_THROTTLE and increase the throttle to something like 60.0.

adrienbarral replied:
And what if I display this message only if obstacle_range > 0 (and < to min_raytrace_range) ? This will handle the particular case of sensor that report 0 when they cannot see.

mikeferguson commented on costmap_2d/plugins/obstacle_layer.cpp:
```diff
@@ -525,6 +531,31 @@ void ObstacleLayer::raytraceFreespace(const Observation& clearing_observation, d
+      }
+      double scale_factor = clearing_observation.start_clearing_range_/obstacle_range;
+      double x_start_clearing = (wx-ox)*scale_factor;
+      double y_start_clearing = (wy-oy)*scale_factor;
```
you aren't actually using these below in the raytrace -- so none of this computation is yet changing anything in your costmap!

mikeferguson replied:
Also, x_start_clearing needs to be added to ox at some point (same for y) to get an actual point. Your worldToMap check is checking the wrong points.

adrienbarral replied:
About your first comment : In the next line, I call
`worldToMap(x_start_clearing, y_start_clearing, x0, y0)`
So this method will update x0, and y0. Both variables are then passed in argument to the method "raytraceLine". So I think that this have an impact in my costmap.

About your second comment : I can't find any emote icon to describe all my shame !!!!

mikeferguson reviewed (changes requested):
I like @procopiostein suggestion that we also implement min_obstacle_range -- I would suggest we go all the way with this and also rename obstacle_range to max_obstacle_range.

mikeferguson commented:
For clarity:

 * Let us define D as the distance from the sensor origin to the measured point (x,y) seen by our sensor.
 * Let us define R as a ray/line from the sensor origin to the measured point (x,y). Note: the ray does not include the actual voxel in which the measured point falls.

And here is how the parameters should probably work: 

 * min_obstacle_range - if D < min_obstacle_range, we will not mark x,y as an obstacle
 * max_obstacle_range (currently just obstacle_range) - if D > max_obstacle_range, we will not mark x,y as an obstacle.
 * min_raytrace_range - if D > min_raytrace_range, then we will start clearing along ray R from a distance min_raytrace_range from the sensor origin.
 * max_raytrace_range (currently just raytrace_range) - if D > max_raytrace_range, then we will stop clearing along ray R at a distance max_raytrace_range from the origin.

Should min_raytrace_range = 0, and the ray R falls entirely within the map_bounds, and D < min_raytrace_range, then we will raytrace all of ray R.

Commit:
Fixing bug in my implementation of start_clearing_range !

Commit:
Renaming raytrace_range into max_raytrace_range.
Renaming start_clearing_range into min_raytrace_range.

Commit:
making maxRaytraceRange works with a non null minRaytraceRange.

Commit:
obstacle_range renamed in max_obstacle_range.

Commit:
Adding min_obstacle_range. If we observe an hit closest than min_obstacle_range, free space will be added, but the lethal obstacle will not be inserted.

Commit:
Fixing typo in a warning.
Using 3D points to compute range when clearing out free spaces. (rather than projection on XYPlane).

mikeferguson commented:
Targets a dead release (kinetic). If this is still of interest, please open a PR to noetic-devel branch
